Cabal/src/Distribution/Simple/GHC/Build/ExtraSources.hs

```diff
@@ -0,0 +1,240 @@
+  -> FilePath
+  -- ^ The build directory for this target
+  -> PreBuildComponentInputs
+  -- ^ The context and component being built in it.
+  -> IO (NubListR FilePath)
+  -- ^ Returns the (nubbed) list of extra sources that were built
+buildAllExtraSources =
+  mconcat
+    [ buildCSources
+    , buildCxxSources
+    , buildJsSources
+    , buildAsmSources
+    , buildCmmSources
+    ]
+
+buildCSources
+  , buildCxxSources
+  , buildJsSources
+  , buildAsmSources
+  , buildCmmSources
+    :: ConfiguredProgram
+    -- ^ The GHC configured program
+    -> FilePath
+    -- ^ The build directory for this target
+    -> PreBuildComponentInputs
+    -- ^ The context and component being built in it.
+    -> IO (NubListR FilePath)
+    -- ^ Returns the list of extra sources that were built
+buildCSources =
+  buildExtraSources
+    "C Sources"
+    Internal.componentCcGhcOptions
+    True
+    ( \c ->
+        cSources (componentBuildInfo c)
+          ++ case c of
+            CExe exe | isC (modulePath exe) -> [modulePath exe]
+            _otherwise -> []
+    )
+buildCxxSources =
+  buildExtraSources
+    "C++ Sources"
+    Internal.componentCxxGhcOptions
+    True
+    ( \c ->
+        cxxSources (componentBuildInfo c)
+          ++ case c of
+            CExe exe | isCxx (modulePath exe) -> [modulePath exe]
+            _otherwise -> []
+    )
+buildJsSources ghcProg buildTargetDir = do
+  Platform hostArch _ <- hostPlatform <$> localBuildInfo
+  let hasJsSupport = hostArch == JavaScript
+  buildExtraSources
+    "JS Sources"
+    Internal.componentJsGhcOptions
+    False
+    ( \c ->
+        if hasJsSupport
+          then -- JS files are C-like with GHC's JS backend: they are
+          -- "compiled" into `.o` files (renamed with a header).
+          -- This is a difference from GHCJS, for which we only
+          -- pass the JS files at link time.
+            jsSources (componentBuildInfo c)
+          else mempty
+    )
+    ghcProg
+    buildTargetDir
+buildAsmSources =
+  buildExtraSources
+    "Assembler Sources"
+    Internal.componentAsmGhcOptions
+    True
+    (asmSources . componentBuildInfo)
+buildCmmSources =
+  buildExtraSources
+    "C-- Sources"
+    Internal.componentCmmGhcOptions
+    True
+    (cmmSources . componentBuildInfo)
+
+-- | Create 'PreBuildComponentRules' for a given type of extra build sources
+-- which are compiled via a GHC invocation with the given options. Used to
+-- define built-in extra sources, such as, C, Cxx, Js, Asm, and Cmm sources.
+buildExtraSources
+  :: String
+  -- ^ String describing the extra sources being built, for printing.
+  -> (Verbosity -> LocalBuildInfo -> BuildInfo -> ComponentLocalBuildInfo -> FilePath -> FilePath -> GhcOptions)
+  -- ^ Function to determine the @'GhcOptions'@ for the
+  -- invocation of GHC when compiling these extra sources (e.g.
+  -- @'Internal.componentCxxGhcOptions'@,
+  -- @'Internal.componentCmmGhcOptions'@)
+  -> Bool
+  -- ^ Want dynamic?
```

Comment:
Who is wanting dynamic linking (I assume)? Is it that some kind extra sources need to force dynamic linking? Actually it seem to only affect `ghcOptFPic`, I am not sure. Can you explain in the comment the role of this. 
Can we also use a more descriptive rather than a Bool?

Reply:
I think this is about JS sources not requiring particularly to be built in the Dynamic way, even if the GHC is dynamic, because they behave differently from other object files like those from C. It may be as well because GHCJS could not even support dynamic-related flags such as -fPIC.

Really, this was the way it was done before: all extra sources except for GHCJS do "wantDyn = True", and JS sources (for the GHCJS compiler) "wantDyn = False". So I'm just preserving this behaviour.

I could investigate, but I think it would be better not mess around with GHCJS bits, especially because GHC's JS backend behaves somewhat differently to GHCJS and might result in a more uniform story here.

I will improve the comment.